### PR TITLE
Fix some Contains subqueries failures with Linq

### PR DESCRIPTION
**Pull request by gliljas**
This is most likely an old regression, where the smartness around LINQ Contains queries ignores the fact that source to the Contains can contain body clauses. In my case it was the use of `MappedAs` that was left hanging in the querymodel, even though `removeMappedAsCalls` was set to true, but I guess there may be other issues.

### CODE REVIEW

**Comment by fredericDelaporte** [comment]
It was not clear for me if this was an improvement or a bug fix. Checking the test case, the query throws without the change. So, labeling as bug.

**Comment by gliljas** [comment]
Yes, it's a bug.

**Comment by fredericDelaporte** [comment]
Though not a recent regression, should we target 5.4.1 instead of master?

**Review by fredericDelaporte — APPROVED** [review]

File: src/NHibernate/Linq/Visitors/ParameterTypeLocator.cs [hunk 1]
```diff
@@ -302,6 +302,7 @@ private bool TryLinkContainsMethod(QueryModel queryModel)
 				// ContainsResultOperator where the constant expression is dislocated from the related expression,
 				// we have to manually link the related expressions.
 				if (queryModel.ResultOperators.Count != 1 ||
+					queryModel.BodyClauses.Count > 0 ||
```
Comment (bahusoid) [inline src/NHibernate/Linq/Visitors/ParameterTypeLocator.cs:305]: We also have different issue regarding `Contains` on parameter constants. See #2872. In 5.3 we didn't handle it properly but exception is thrown.  
I wonder if this change makes `Where` part be silently ignored.

Comment (gliljas) [reply src/NHibernate/Linq/Visitors/ParameterTypeLocator.cs:305]: I must admit I don't quite understand why TryLinkContainsMethod is needed.

Comment (bahusoid) [reply src/NHibernate/Linq/Visitors/ParameterTypeLocator.cs:305]: It's required for proper parameter type recognition(See #2459). So it's required for queries like `Where(order => collectionParameter.Contains(order.Id))` or `Where(order => order.Lines.Contains(parameter))`
So in theory your changes might break second type of queries (Contains on subquery) with Body clauses. But removing completely this check I see fails only for first type of queries (Contains on collection parameter). So it's either not properly covered in tests or this change not needed any more...

Comment (bahusoid) [reply src/NHibernate/Linq/Visitors/ParameterTypeLocator.cs:305]: > So in theory your changes might break second type of queries

So maybe to be on a safe side we should keep logic adding related expressions and instead replace final "return true" that skips further processing with something like:
```C#
//Subqueries not based on parameters need further processing
return right.NodeType != ExpressionType.Constant;
```

Comment (bahusoid) [reply src/NHibernate/Linq/Visitors/ParameterTypeLocator.cs:305]: My last suggestion is actually wrong. And it should be actually pretty safe as is... So marking it as resolved.

**Comment by bahusoid** [comment]
Replaced by #3274

**Pull request closed without merging**